### PR TITLE
Improve picchu reconcile performance and superfluous resource updates

### PR DESCRIPTION
**Pull request by dokipen**

### CODE REVIEW

**Comment by dokipen** [comment]
:zap:

File: pkg/controller/releasemanager/plan/syncApp.go [hunk 1]
```diff
@@ -445,6 +445,10 @@ func (p *SyncApp) service() *corev1.Service {
 		labels[k] = v
 	}
 
+	// Don't needlessly update services
```
Comment (silverlyra) [inline pkg/controller/releasemanager/plan/syncApp.go:448]: 🙏

File: pkg/controller/releasemanager/releasemanager_controller.go [hunk 2]
```diff
@@ -152,8 +153,9 @@ func newReconciler(mgr manager.Manager, c utils.Config) reconcile.Reconciler {
 }
 
 // add adds a new Controller to mgr with r as the reconcile.Reconciler
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
+func add(mgr manager.Manager, r reconcile.Reconciler, c utils.Config) error {
 	_, err := builder.ControllerManagedBy(mgr).
+		WithOptions(controller.Options{MaxConcurrentReconciles: c.ConcurrentReleaseManagers}).
```
Comment (silverlyra) [inline pkg/controller/releasemanager/releasemanager_controller.go:158]: :mind-blown:

File: pkg/controller/revision/revision_controller.go [hunk 3]
```diff
@@ -256,12 +260,30 @@ func (r *ReconcileRevision) Reconcile(request reconcile.Request) (reconcile.Resu
...
 
+	return r.Requeue(log, nil)
+}
+
+func (r *ReconcileRevision) Requeue(log logr.Logger, err error) (reconcile.Result, error) {
+	if err != nil {
+		log.Error(err, "Reconcile resulted in error")
+		return reconcile.Result{}, err
+	}
+	log.Info("Reconciled successfully", "requeueAfter", r.config.RequeueAfter)
```
Comment (silverlyra) [inline pkg/controller/revision/revision_controller.go:274]:
```suggestion
	log.Info("Reconciled successfully", "requeue", true, "requeueAfter", r.config.RequeueAfter)
```

File: pkg/controller/revision/revision_controller.go [hunk 4]
```diff
@@ -256,12 +260,30 @@ func (r *ReconcileRevision) Reconcile(request reconcile.Request) (reconcile.Resu
...
 	return reconcile.Result{RequeueAfter: r.config.RequeueAfter}, nil
 }
 
+func (r *ReconcileRevision) NoRequeue(log logr.Logger, err error) (reconcile.Result, error) {
+	if err != nil {
+		log.Error(err, "Reconcile resulted in error")
+		return reconcile.Result{}, err
+	}
+	log.Info("Reconciled successfully, no requeue")
```
Comment (silverlyra) [inline pkg/controller/revision/revision_controller.go:283]:
```suggestion
	log.Info("Reconciled successfully", "requeue", false)
```

will make these logs more searchable

File: pkg/plan/common.go [hunk 5]
```diff
@@ -2,24 +2,39 @@ package plan
...
 	"k8s.io/apimachinery/pkg/runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
+type NoChangeNeededError error
+
+func NoChangeNeeded() error {
+	return NoChangeNeededError(errors.New("no change needed"))
+}
```
Comment (silverlyra) [inline pkg/plan/common.go:28]:
```suggestion
var ErrNoChangeNeeded = NoChangeNeededError(errors.New("no change needed"))
```

this better fits the Go convention around distinguished errors

File: pkg/plan/common.go [hunk 6]
```diff
@@ -312,18 +327,45 @@ func CreateOrUpdate(
...
 			}
+			// Only update replicas if we are changing to/from zero, which means the replicaset is being retired/deployed
 			var zero int32 = 0
 			if rs.Spec.Replicas == nil {
 				rs.Spec.Replicas = &zero
 			}
 			if 0 == *typed.Spec.Replicas || 0 == *rs.Spec.Replicas {
 				*rs.Spec.Replicas = *typed.Spec.Replicas
+			} else {
+				// if the only change is replicas, and neither is zero, abort update. Replicas can change so fast in
```
Comment (matkam) [inline pkg/plan/common.go:338]: Will this work with WPAs that can scale ReplicaSet to 0?

File: pkg/plan/common.go [hunk 7]
```diff
@@ -2,24 +2,39 @@ package plan
...
+}
+
+func IsNoChangeNeededError(err error) bool {
+	switch err.(type) {
+	case NoChangeNeededError:
+		return true
+	}
+	return false
+}
+
```
Comment (silverlyra) [inline pkg/plan/common.go:37]:
```suggestion
```

File: pkg/plan/common.go [hunk 8]
```diff
@@ -312,18 +327,45 @@ func CreateOrUpdate(
...
 			}
 			if 0 == *typed.Spec.Replicas || 0 == *rs.Spec.Replicas {
 				*rs.Spec.Replicas = *typed.Spec.Replicas
+			} else {
+				// if the only change is replicas, and neither is zero, abort update. Replicas can change so fast in
+				// response to HPAs that we often get "fake" changes
+				obj := typed.DeepCopy()
+				obj.Spec.Replicas = rs.Spec.Replicas
+				if equality.Semantic.DeepEqual(rs, obj) {
+					return NoChangeNeeded()
```
Comment (silverlyra) [inline pkg/plan/common.go:343]:
```suggestion
					return ErrNoChangeNeeded
```

File: pkg/plan/common.go [hunk 9]
```diff
@@ -312,18 +327,45 @@ func CreateOrUpdate(
...
+						updateLabels = true
+					}
+				}
+			}
+			if updateLabels {
+				rs.Labels = CopyStringMap(typed.Labels)
+			}
 			return nil
 		})
+		if IsNoChangeNeededError(err) {
```
Comment (silverlyra) [inline pkg/plan/common.go:365]:
```suggestion
		if errors.Is(err, ErrNoChangeNeeded) {
```

File: pkg/plan/common.go (hunk 6 above)
Comment (dokipen) [reply pkg/plan/common.go:338]: I don't think this PR makes a difference in that regard. Existing picchu will keep overwriting `replicas: 0` with default replica count if the revision is deployed as is.

File: pkg/plan/common.go (hunk 9 above)
Comment (dokipen) [reply pkg/plan/common.go:365]: TIL

File: pkg/controller/releasemanager/releasemanager_controller.go (hunk 2 above)
Comment (dokipen) [reply pkg/controller/releasemanager/releasemanager_controller.go:158]: btw, the limiting factor on revision concurrency is DDOSing prom endpoint.

**Pull request merged**